File: 2018/python/day09.py
from collections import deque

def play(players, marbles):
    scores = [0] * players
    Q = deque([0])
    for m in range(1, marbles + 1):
        if m % 23 == 0:
            Q.rotate(7)
            scores[m%players] += Q.pop() + m
            Q.rotate(-1)
        else:
            Q.rotate(-1)
            Q.append(m)
    return (max(scores))

input = list(open("../input/day9.input").readline().strip("\n").split(" "))
players, marbles = int(input[0]), int(input[6]) 
print("day 9, part a:", play(players, marbles))
print("day 9, part b:", play(players, marbles*100))

File: 2018/csharp/day09.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace aoc2018
{
    class Day9
    {
        public static void Solve()
        {
            var input = File.ReadAllText(Path.Combine("..", "input", "day9.input")).Split(" ");
            int players = int.Parse(input[0]);
            int marbles = int.Parse(input[6]) * 100;
            var circle = new LinkedList<int>();
            circle.AddFirst(0);
            var current = circle.First;
            var scores = new Dictionary<int, long>();
            Enumerable.Range(0, players).ToList().ForEach(i => scores[i] = 0);

            for (int i = 1; i < marbles; i++)
            {
                if (i == marbles / 100)
                    Console.WriteLine("day 9, answer a: {0}", scores.Values.Max());
                if (i % 23 == 0)
                {
                    for (int j = 0; j < 7; j++)
                        current = current.Previous ?? circle.Last;
                    scores[i % players] += current.Value;
                    scores[i % players] += i;
                    var next = current.Next;
                    circle.Remove(current);
                    current = next;
                }
                else
                    current = current.Next == null ? circle.AddAfter(circle.First, i) : circle.AddAfter(current.Next, i);
            }
            Console.WriteLine("day 9, answer b: {0}", scores.Values.Max());
        }
    }
}
